File: src/test/java/fi/metatavu/oioi/cm/test/functional/builder/TestBuilder.java
package fi.metatavu.oioi.cm.test.functional.builder;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Predicate;
import java.util.stream.Collectors;

import fi.metatavu.oioi.cm.test.functional.builder.auth.DefaultAccessTokenProvider;
import fi.metatavu.oioi.cm.test.functional.builder.auth.InvalidAccessTokenProvider;
import fi.metatavu.oioi.cm.test.functional.builder.auth.NullAccessTokenProvider;
import fi.metatavu.oioi.cm.test.functional.builder.auth.TestBuilderAuthentication;

/**
 * Test builder class
 * 
 * @author Antti Leppä
 */
public class TestBuilder implements AutoCloseable {

  private static final String REALM = "oioi";
  private static final String CLIENT_ID = "ui";
  private static final String ADMIN_USER = "admin";
  private static final String ADMIN_PASSWORD = "admin";
  private static final String CLIENT_SECRET = null;

  private TestBuilderAuthentication admin;
  private TestBuilderAuthentication invalid;
  private TestBuilderAuthentication anonymous;
  private List<CloseableResource<?, ?>> closables = new ArrayList<>();

  /**
   * Returns admin authenticated authentication resource
   * 
   * @return admin authenticated authentication resource
   * @throws IOException 
   */
  public TestBuilderAuthentication admin() throws IOException {
    if (admin != null) {
      return admin;
    }
    
    admin = new TestBuilderAuthentication(this, new DefaultAccessTokenProvider(REALM, CLIENT_ID, ADMIN_USER, ADMIN_PASSWORD, CLIENT_SECRET));
    
    return admin;
  }

  /**
   * Returns authentication resource with invalid token
   * 
   * @return authentication resource with invalid token
   */
  public TestBuilderAuthentication invalid() {
    if (invalid != null) {
      return invalid;
    }

    return invalid = new TestBuilderAuthentication(this, new InvalidAccessTokenProvider());
  }

  /**
   * Returns authentication resource without token
   * 
   * @return authentication resource without token
   */
  public TestBuilderAuthentication anonymous() {
    if (anonymous != null) {
      return anonymous;
    }

    return anonymous = new TestBuilderAuthentication(this, new NullAccessTokenProvider());
  }

  /**
   * Adds closable to clean queue
   * 
   * @param closable closable
   * @return given instance
   */
  protected <T extends CloseableResource<?, ?>> T addClosable(T closable) {
    closables.add(closable);
    return closable;
  }

  /**
   * Removes a closable from clean queue
   * 
   * @param predicate filter predicate
   */
  protected <T> void removeCloseable(Predicate<Object> predicate) {
    closables = closables.stream().filter((closeable) -> {
      return !predicate.test(closeable.getResource());
    }).collect(Collectors.toList());
  }

  @Override
  public void close() throws Exception {
    for (int i = closables.size() - 1; i >= 0; i--) {
      closables.get(i).close();
    }
  }

}

File: src/test/kotlin/fi/metatavu/oioi/cm/test/functional/builder/TestBuilder.kt
package fi.metatavu.oioi.cm.test.functional.builder

import fi.metatavu.jaxrs.test.functional.builder.AbstractTestBuilder
import fi.metatavu.jaxrs.test.functional.builder.auth.AccessTokenProvider
import fi.metatavu.jaxrs.test.functional.builder.auth.AuthorizedTestBuilderAuthentication
import fi.metatavu.jaxrs.test.functional.builder.auth.KeycloakAccessTokenProvider
import fi.metatavu.oioi.cm.client.infrastructure.ApiClient
import fi.metatavu.oioi.cm.test.functional.builder.auth.TestBuilderAuthentication
import org.eclipse.microprofile.config.ConfigProvider
import java.io.IOException

/**
 * TestBuilder implementation
 *
 * @author Antti Leppä
 * @author Jari Nykänen
 */
open class TestBuilder: AbstractTestBuilder<ApiClient> () {

    val settings = ApiTestSettings()
    val admin = createTestBuilderAuthentication(username = "admin", password = "admin")
    val customer1User = createTestBuilderAuthentication(username = "customer-1-user", password = "pass")
    val customer1Admin = createTestBuilderAuthentication(username = "customer-1-admin", password = "pass")
    val customer2User = createTestBuilderAuthentication(username = "customer-2-user", password = "pass")
    val customer2Admin = createTestBuilderAuthentication(username = "customer-2-admin", password = "pass")

    override fun createTestBuilderAuthentication(testBuilder: AbstractTestBuilder<ApiClient>, accessTokenProvider: AccessTokenProvider): AuthorizedTestBuilderAuthentication<ApiClient> {
        return TestBuilderAuthentication(this, accessTokenProvider)
    }

    /**
     * Creates test builder authenticatior for given user
     *
     * @param username username
     * @param password password
     * @return test builder authenticatior for given user
     */
    private fun createTestBuilderAuthentication(username: String, password: String): TestBuilderAuthentication {
        val authServerUrl: String = ConfigProvider.getConfig().getValue("oioi.keycloak.url", String::class.java)
        val realm: String = ConfigProvider.getConfig().getValue("oioi.keycloak.realm", String::class.java)
        val clientId = "ui"
        return TestBuilderAuthentication(this, KeycloakAccessTokenProvider(authServerUrl, realm, clientId, username, password, null))
    }
    
}